Reduce binary size of query_builder.cpp and query.cpp

RedBeard0531 opened the pull request:
480k reduction on realm-java arm64 build with parser

* pull throwing of exceptions out to non-templated functions
* convert templated functions in query_builder.cpp to return Query so the code
  to add it to the tree is only needed once.
* FORCEINLINE overloads for invalid combinations
  * Counterintutively, this actually shrinks the binary because the compiler
    is able to directly call the `throw_foo_error` function, which propagates
    the information about it never returning. It is also able to redirect
    branches that lead there to the same instructions, which it wasn't able to
    do previously because each template instantiation was considered a
    different function, even though they did the same thing.

RedBeard0531 commented:
@ironage I know the function names no longer make sense and will be renamed prior to pushing. But I thought it would be easier to review before they all get changed.

ironage commented on src/realm/parser/query_builder.cpp:
```diff
@@ -36,131 +36,129 @@ using namespace query_builder;
+{
     switch (operatorType) {
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
```
We may not be tracking these types of warnings right now, but I at least found this macro to be a useful indication that the the fallthrough is not a programmatic error. Do you have a reason for removing it?

RedBeard0531 replied:
I think the warning is on by default with -Wall. However you only need to annotate when there is code between the cases, not when cases are directly stacked like this. I kinda like having the fall through annotation be a signal that there is "interesting" control flow going on, not just multiple cases handled by identical code. 

That said, if you found them helpful in these cases, I can put them back.

ironage replied:
> However you only need to annotate when there is code between the cases, not when cases are directly stacked like this.

I must have forgotten this. There is no style convention for this that I know of and your reasoning makes sense to me, so I'm fine with the removal in the trivial cases. Thanks for explaining!

RedBeard0531 replied:
BTW, while it isn't obvious due to the trivial cases like this, this patch *does* remove some cases of "interesting" fallthrough (not just removing the macro 😉), such as in `do_make_null_comparison_query` and ` make_link_constraint_query`. It is probably worth reviewing those a bit more closely.

RedBeard0531 commented:
@ironage I've rebased to the top of the v10 branch, renamed the incorrectly named functions, added a few more size optimizations, and addressed compiler warnings in order to get a green build. It should now be ready for a proper review.

I recommend either reviewing the individual commits or at least reading the commit messages, since they are topical and include explanations. Before merging, I'll squash these down to one or a few clean commits and get rid of any fixup commits.

ironage commented on src/realm/parser/query_builder.cpp:
```diff
@@ -37,131 +37,129 @@ using namespace query_builder;
-        case Predicate::Operator::NotEqual:
-            query.Not();
-            REALM_FALLTHROUGH;
+        case Predicate::Operator::NotEqual: {
```
This is certainly a win in terms of readability. It might have been intending to use a potentially more optimized path for equality queries and inverting the result via "Not()" rather than using the NotEqual operator. If it doesn't show up in the performance report then it isn't something we have explicitly optimized for and I'd be fine with the change. The problem is that it appears the performance report is not displaying the results of this PR due to a bug in the performance scripts. I'll see if I can fix that so we can view the results.

ironage replied:
The performance report [here](https://ci.realm.io/job/realm/job/realm-core/job/PR-3858/3/Performance_5fReport/) from https://github.com/realm/realm-core/pull/3858 indicates that we are either not explicitly tracking this code path, or it is has negligible performance impact so I'm satisfied with that.

RedBeard0531 commented:
> Do you have a good way to track the size change?

Unfortunately not really. I was using a fork of realm-java that added in the parser functionality.

> I'd like to add that to the performance report if possible. We are tracking the size of librealm.a and I could add tracking for librealm-parser.a as well, but I'm not sure that is the metric to track? 

The `.a` files are not a good thing to track, since they are just an archive of `.o` files, so they haven't undergone the normal linking process which doesn't things like deduping COMDAT symbols (ie stuff in headers that is emitted in each `.o` that uses it). And it definitely can't take advantage of things like LTO, `-gc-sections` (removing stuff not reachable from entry points), or `--icf=all` (identical code folding). The `.so` files would be closer to a good thing to track, but when I tried that with LTO (which hopefully the SDKs that care about either size or speed will switch to soon), the parser lib was unreasonably small (~38k IIRC). I think the linker realized that almost all of the code was unused because of hidden linkage.

In order to get an accurate metric, we would need to link librealm-core and librealm-parser together with a stub that consumes all of the "public" API used by the SDKs. But at that point it is probably easier and better to just directly use an SDK (or many/all of them). This would be one advantage of going "full monorepo" 😄
